source/common/http/http2/codec_impl.cc

```diff
@@ -48,8 +48,8 @@ template <typename T> static T* remove_const(const void* object) {
 }
 
 ConnectionImpl::StreamImpl::StreamImpl(ConnectionImpl& parent, uint32_t buffer_limit)
-    : parent_(parent), headers_(new HeaderMapImpl()), local_end_stream_sent_(false),
-      remote_end_stream_(false), data_deferred_(false),
+    : parent_(parent), headers_(new HeaderMapImpl()), metadata_map_(new MetadataMap()),
```

Comment:
What do we think about doing lazy creation here?  I think it's worth the complexity to not create a map for the vast majority of users who don't need this functionality.

Also if there's one metadata_map_ (at a time) per stream and one decoder per stream I'm not sure I follow why you moved ownership out of the decoder.

Reply:
Right! Moved metadata_map_ back to decoder. I thought it would be consistent to let StreamImpl hold metadata since it holds headers. But it is definitely not necessary.